src/Common/Core/Impl/Services/ServiceManager.cs

```diff
@@ -0,0 +1,140 @@
+using Microsoft.Common.Core.Diagnostics;
+
+namespace Microsoft.Common.Core.Services {
+    public class ServiceManager : IServiceManager {
+        private readonly Dictionary<Type, object> _services = new Dictionary<Type, object>();
+        private readonly Dictionary<Type, Func<object>> _deferredServices = new Dictionary<Type, Func<object>>();
+        private readonly object _lock = new object();
+
+        /// <summary>
+        /// Fire when service is added
+        /// </summary>
+        public event EventHandler<ServiceContainerEventArgs> ServiceAdded;
+        /// <summary>
+        /// Fires when service is removed
+        /// </summary>
+        public event EventHandler<ServiceContainerEventArgs> ServiceRemoved;
+
+        /// <summary>
+        /// Add service to the service manager container
+        /// </summary>
+        /// <typeparam name="T">Service type</typeparam>
+        /// <param name="service">Service instance</param>
+        public virtual IServiceManager AddService<T>(T service) where T : class {
+            var type = typeof(T);
+            Check.ArgumentNull(nameof(service), service);
+            lock (_lock) {
+                Check.InvalidOperation(() => _services.ContainsKey(type));
+                _services[type] = service;
+            }
+            ServiceAdded?.Invoke(this, new ServiceContainerEventArgs(type));
+            return this;
+        }
+
+        /// <summary>
+        /// Adds on-demand created service
+        /// </summary>
+        /// <param name="type">Service type</param>
+        /// <param name="factory">Optional creator function. If not provided, reflection with default constructor will be used.</param>
+        /// <param name="parameters">Factory parameters</param>
+        public virtual IServiceManager AddService(Type type, Func<object> factory) {
+            Check.ArgumentNull(nameof(type), type);
+            lock (_lock) {
+                Check.InvalidOperation(() => _services.ContainsKey(type));
+                Check.InvalidOperation(() => _deferredServices.ContainsKey(type));
+                _deferredServices[type] = factory;
+            }
+
+            ServiceAdded?.Invoke(this, new ServiceContainerEventArgs(type));
+            return this;
+        }
+
+        /// <summary>
+        /// Retrieves service from the container
+        /// </summary>
+        /// <typeparam name="T">Service type</typeparam>
+        /// <returns>Service instance or null if it doesn't exist</returns>
+        public virtual T GetService<T>(Type type = null) where T : class {
+            type = type ?? typeof(T);
+
+            lock (_lock) {
+                object service;
+                if (_services.TryGetValue(type, out service)) {
+                    return service as T;
+                }
+
+                // Try walk through and cast. Perhaps someone is asking for IFoo
+                // that is implemented on class Bar but Bar was added as Bar, not as IFoo
```

Comment:
> Perhaps someone is asking for IFoo that is implemented on class Bar but Bar was added as Bar, not as IFoo

So the code that adds a service can't actually specify the type for registration. Not sure that I like it. First, based on the order in which dictionary will enumerate its values, this code can return different instances if there is more than one object that implements requested interface. Second, it doesn't match the logic in the `AddService` method that doesn't check for implemented interfaces.